### PR TITLE
PLFM-7302: Map subject to principal alias when using an oauth provider

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/GoogleOAuth2Provider.java [hunk 1]
```diff
@@ -26,26 +26,25 @@
  */
 public class GoogleOAuth2Provider implements OAuthProviderBinding {
 
-    private static final String AUTHORIZE_URL = "https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=%s&redirect_uri=%s&prompt=select_account";
-    private static final String TOKEN_URL = "https://accounts.google.com/o/oauth2/token";
+    private static final String AUTHORIZE_URL = "https://accounts.google.com/o/oauth2/v2/auth?response_type=code&client_id=%s&redirect_uri=%s&prompt=select_account";
+    private static final String TOKEN_URL = "https://oauth2.googleapis.com/token";
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/GoogleOAuth2Provider.java]: Consider getting `AUTHORIZE_URL`, `TOKEN_URL`, and `GOOGLE_OAUTH_USER_INFO_API_URL` from the Google OIDC discovery document:  from https://accounts.google.com/.well-known/openid-configuration

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/GoogleOAuth2Provider.java [hunk 2]
```diff
@@ -26,26 +26,25 @@
...
-	private static final String GOOGLE_OAUTH_USER_INFO_API_URL = "https://www.googleapis.com/oauth2/v2/userinfo";
+	private static final String GOOGLE_OAUTH_USER_INFO_API_URL = "https://openidconnect.googleapis.com/v1/userinfo";
 	/*
-	 * Json keys for object returned by https://www.googleapis.com/oauth2/v2/userinfo
+	 * Json keys for object returned by https://openidconnect.googleapis.com/v1/userinfo
 	 */
 	public static final String EMAIL = "email";
-	public static final String VERIFIED_EMAIL = "verified_email";
-	public static final String ID = "id";
+	public static final String VERIFIED_EMAIL = "email_verified";
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/oauth/GoogleOAuth2Provider.java]: VERIFIED_EMAIL -> VERIFIED_EMAIL

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java [hunk 3]
```diff
@@ -247,6 +266,16 @@ public PrincipalAlias lookupUserByUsernameOrEmail(String alias) {
 		}
 		return pa;
 	}
+	
+	@Override
+	public PrincipalAlias lookupUserByOauthProvider(OAuthProvider provider, String subject) {
+		String oauthAlias = getAliasForOauthProvider(provider, subject);
+		PrincipalAlias pa = this.principalAliasDAO.findPrincipalWithAlias(getAliasForOauthProvider(provider, subject), AliasType.USER_OAUTH);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java]: PrincipalAlias pa = this.principalAliasDAO.findPrincipalWithAlias(oauthAlias, AliasType.USER_OAUTH);

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java [hunk 4]
```diff
@@ -115,10 +120,24 @@ public long createUser(NewUser user) {
...
+	
+	@WriteTransaction
+	@Override
+	public PrincipalAlias bindOauthProviderAlias(OAuthProvider provider, String subject, Long principalId) {
+		String oauthAlias = getAliasForOauthProvider(provider, subject);
+		return bindAlias(oauthAlias, AliasType.USER_OAUTH, principalId);
+	}
+	
+	private static String getAliasForOauthProvider(OAuthProvider provider, String subject) {
+		return provider.name() + " " + subject;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java]: Note that the 'alias unique' shouldn't change the OIDC alias *at all*.

**Comment by brucehoff** [comment]
TODO:  Add regex to validate provider+subject alias.
TODO:  Add integration test that persists a provider+subject alias.

**Review by brucehoff — CHANGES_REQUESTED** [review]

File: lib/jdomodels/src/main/resources/schema/PrincipalOIDCBinding-ddl.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS `PRINCIPAL_OIDC_BINDING` (
+  `ID` BIGINT NOT NULL,
+  `CREATED_ON` TIMESTAMP(3) NOT NULL,
+  `PRINCIPAL_ID` BIGINT NOT NULL,
+  `PROVIDER` ENUM('GOOGLE_OAUTH_2_0') NOT NULL,
+  `SUBJECT` VARCHAR(256) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL,
```
Comment (john-hill) [inline lib/jdomodels/src/main/resources/schema/PrincipalOIDCBinding-ddl.sql]: the spec says ascii.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/principal/PrincipalOIDCBindingDaoImplTest.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	@Test
+	public void testBindPrincipalToSubjectWithExisting() {
+		
+		String subject = "subject";
+		
+		dao.bindPrincipalToSubject(principalId, OAuthProvider.GOOGLE_OAUTH_2_0, subject);
+		
+		assertThrows(IllegalArgumentException.class, () -> {
+			// Call under test
+			dao.bindPrincipalToSubject(principalId, OAuthProvider.GOOGLE_OAUTH_2_0, subject);
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/principal/PrincipalOIDCBindingDaoImplTest.java]: this should be noop

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java [hunk 7]
```diff
@@ -80,27 +87,29 @@ public long createUser(NewUser user) {
 		if (alias != null) {
 			throw new NameConflictException("User '" + user.getUserName() + "' already exists");
 		}
+		// Make sure that the subject for an oauth provider is not bound yet
+		if (user.getOauthProvider() != null) {
+			lookupUserIdByOIDCSubject(user.getOauthProvider(), user.getSubject()).ifPresent((principalId)-> {
+				throw new NameConflictException("A user is already bound to the '" + user.getOauthProvider() +"' provider with the same identity");
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserManagerImpl.java]: Have you already used this 'google' account to log into another Synapse account?

File: services/repository-managers/src/main/resources/private/managers-spb.xml [hunk 8]
```diff
@@ -79,11 +79,23 @@
 	<bean id="googleOAuth2Provider" class="org.sagebionetworks.repo.manager.oauth.GoogleOAuth2Provider">
 		<constructor-arg ref="stackConfiguration.oAuth2GoogleClientId" index="0" />
 		<constructor-arg ref="stackConfiguration.oAuth2GoogleClientSecret" index="1" />
+		<constructor-arg>
```
Comment (john-hill) [inline services/repository-managers/src/main/resources/private/managers-spb.xml]: can we move out of the xml, using annotations.

**Review by john-hill — CHANGES_REQUESTED** [review]

**Review by john-hill — APPROVED** [review]

**Pull request merged**